tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-InProc.cpp

```diff
@@ -0,0 +1,969 @@
+      case CustomBufferKind::CodeCompletionResultsArray:
+        return SOURCEKITD_VARIANT_TYPE_ARRAY;
+    }
+    llvm::report_fatal_error("sourcekitd object did not resolve to a known type");
+  }
+
+  SKDCustomData *asCustomData() override {
+    return this;
+  }
+
+  CustomBufferKind getKind() {
+    return Kind;
+  }
+
+  const void *getDataPtr() {
+    return BufferPtr->getBuffer().data();
+  }
+
+private:
+  CustomBufferKind Kind;
+  std::unique_ptr<llvm::MemoryBuffer> BufferPtr;
+};
+
+class SKDError: public SKDObject {
+public:
+  SKDError(sourcekitd_error_t Kind, StringRef Description)
+  : Kind(Kind), Description(Description.str()) {}
+
+  SKDError(SKDError const&) = delete;
+  SKDError &operator=(SKDError const&) = delete;
+
+  sourcekitd_variant_type_t getVariantType() override {
+    llvm::report_fatal_error("sourcekitd error object is not a variant type");
+  }
+
+  SKDError *asError() override {
+    return this;
+  }
+
+  sourcekitd_error_t getKind() {
+    return Kind;
+  }
+
+  StringRef getDescription() {
+    return StringRef(Description);
+  }
+
+private:
+  sourcekitd_error_t Kind;
+  std::string Description;
+};
+
+template <typename ImplClass, typename RetTy = void>
+class SKDObjectVisitor {
+public:
+  typedef std::vector<std::pair<UIdent, sourcekitd_object_t>> DictMap;
+
+  static bool compKeys(const std::pair<UIdent, sourcekitd_object_t> &LHS,
+                       const std::pair<UIdent, sourcekitd_object_t> &RHS) {
+    return sourcekitd::compareDictKeys(LHS.first, RHS.first);
+  }
+
+  RetTy visit(sourcekitd_object_t Obj) {
+    auto Object = static_cast<SKDObject *>(Obj);
+
+    if (auto DictionaryObject = Object->asDictionary()) {
+      DictMap Dict;
+      DictionaryObject->apply([&](sourcekitd_uid_t Key, SKDObject *Element){
+        UIdent UID = UIdentFromSKDUID(Key);
+        Dict.push_back({ UID, Element });
+        return true;
+      });
+      std::sort(Dict.begin(), Dict.end(), compKeys);
+      return static_cast<ImplClass*>(this)->visitDictionary(Dict);
+    }
+    if (auto ArrayObject = Object->asArray()) {
+      std::vector<sourcekitd_object_t> Vec;
+      ArrayObject->apply([&](size_t Index, SKDObject *Element){
+        Vec.push_back(Element);
+        return true;
+      });
+      return static_cast<ImplClass*>(this)->visitArray(Vec);
+    }
+    if (auto SKDUID = Object->getUID()) {
+      UIdent UID = UIdentFromSKDUID(SKDUID);
+      return static_cast<ImplClass*>(this)->visitUID(UID.getName());
+    }
+    auto OptString = Object->getString();
+    if (OptString.hasValue()) {
+      return static_cast<ImplClass*>(this)->visitString(StringRef(OptString.getValue()));
+    }
+    auto OptInt = Object->getInt64();
+    if (OptInt.hasValue()) {
+      return static_cast<ImplClass*>(this)->visitInt64(OptInt.getValue());
+    }
+    llvm_unreachable("unknown sourcekitd_object_t");
+  }
+};
+
+class SKDObjectPrinter : public SKDObjectVisitor<SKDObjectPrinter> {
+  raw_ostream &OS;
```

Comment:
Ah, hm, you are right. I think your initial suggestion to move SKDObjectPrinter as common code, which is practically identical, is a good one. We can have a common printer object in a header that is shared between implementations, like this:
```
class SKDObjectPrinterBase {
public:
  SKDObjectPrinter(raw_ostream &OS, unsigned Indent = 0)
  void visitDictionary(const DictMap &Map);
  void visitArray(ArrayRef<sourcekitd_object_t> Arr);
<...>
};
```
And then for each implementation have this:
```
class SKDObjectPrinter : public SKDObjectVisitor<SKDObjectPrinter>, public SKDObjectPrinterBase {
}
````